File: leetcode/UndergroundSystem.py
class UndergroundSystem:

    def __init__(self):
        self.ckIn = {} # self.ckIn:id = [from,time]
        self.ckOut = {} # self.ckOut:(startStation,endStation) = [sum of time , sum of count] 

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        if id in self.ckIn :
            if self.ckIn[id][1] != 0:
                print("error:" , id , self.ckIn[id][0] , self.ckIn[id][1])
                return
            else :
                self.ckIn[id][0] = stationName 
                self.ckIn[id][1] = t
        else :
            self.ckIn[id] = [stationName,t]

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        f = self.ckIn[id][0]
        startTime = self.ckIn[id][1]
        self.ckIn[id][1] = 0
        if (f,stationName) in self.ckOut:
            self.ckOut[(f,stationName)][0] += t-startTime
            self.ckOut[(f,stationName)][1] += 1
        else:
            self.ckOut[(f,stationName)] = [t-startTime,1]
        

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        if (startStation,endStation) in self.ckOut:
            return self.ckOut[(startStation,endStation)][0] / self.ckOut[(startStation,endStation)][1]
        else :
            return 0.0


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.self.ckIn(id,stationName,t)
# obj.self.ckOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

File: leetcode/2023/UndergroundSystem.cpp
#include <map>
#include <string>



using namespace std;

class UndergroundSystem {
public:
    map<int,pair<string,int>> checkin;
    map<pair<string,string>,pair<int,int>> checkout;
    UndergroundSystem() {
        
    }
    
    void checkIn(int id, string stationName, int t) {
        map<int,pair<string,int>>::iterator f = checkin.find(id);
        if( f == checkin.end()){
            checkin.insert(make_pair(id,make_pair(stationName,t)));
        } else {
            checkin[id].first = stationName;
            checkin[id].second = t;
            //  f->second = t;   it has error : const int can not modify.
            // f->second is pair<sring,int>
            // f->first = id
        }
    }
    
    void checkOut(int id, string stationName, int t) {
        auto p = make_pair(checkin[id].first,stationName);
        auto itco = checkout.find(make_pair(checkin[id].first,stationName));
        if( itco == checkout.end()){
            checkout.insert(make_pair(p,make_pair(t - checkin[id].second,1)));
        } else {
            checkout[p].first += t - checkin[id].second;
            checkout[p].second ++;
        }
    }
    
    double getAverageTime(string startStation, string endStation) {
        auto p = make_pair(startStation,endStation);
        // cout << startStation << " ";
        // cout << endStation << " ";
        // cout << checkout[p].first << " ";
        // cout << checkout[p].second << " ";
        // printf("%f\n",(double) (1.0 * checkout[p].first / checkout[p].second));
        return (1.0 * checkout[p].first / checkout[p].second);
    }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */